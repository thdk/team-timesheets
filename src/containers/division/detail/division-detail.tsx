import React, { useCallback } from "react";
import { TextField } from "@rmwc/textfield";

import { Form, FormField } from "../../../components/layout/form";
import { FlexGroup } from "../../../components/layout/flex";
import IconPicker from "../../../components/icon-picker";

type Props = {
    readonly onNameChanged: (name: string) => void;
    readonly onIconChanged: (icon: string) => void;
    readonly icon?: string;
    readonly name?: string;
}

export const DivisionDetail = (props: Props) => {
    const {
        icon,
        name,
        onIconChanged,
        onNameChanged,
    } = props;

    const handleNameChanged = useCallback((e: React.ChangeEvent<HTMLInputElement>) => {
        onNameChanged(e.target.value);
    }, [onNameChanged]);

    const handleIconChanged = useCallback((icon: string) => {
        onIconChanged(icon);
    }, [onIconChanged]);

    return <>
        <Form>
            <FlexGroup extraCssClass="row">
                <FormField>
                    <TextField
                        outlined
                        label="Name"
                        value={name}
                        onChange={handleNameChanged}
                    />
                </FormField>
                <FormField first={false}>
                    <IconPicker
                        onChange={handleIconChanged}
                        value={icon}
                        outlined
                    />
                </FormField>
            </FlexGroup>
        </Form>
    </>;
};
DivisionDetail.displayName = "DivisionDetail";