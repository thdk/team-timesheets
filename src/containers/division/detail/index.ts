export { DivisionDetailContainer as DivisionDetail } from "./division-detail-container";
