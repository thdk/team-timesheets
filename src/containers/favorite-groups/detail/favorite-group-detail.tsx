import React, { useState, useCallback } from "react";
import { Form, FormField } from "../../../components/layout/form";
import { FlexGroup } from "../../../components/layout/flex";
import { TextField } from "@rmwc/textfield";
import { IFavoriteRegistrationGroup } from "../../../../common";
import { FavoriteGroupSelect } from "../select";
import { Checkbox } from "@rmwc/checkbox";

export const FavoriteGroupDetailForm = ({
    groups,
    group: {
        name,
        id,
    },
    onNameChanged,
}: {
    groups: (IFavoriteRegistrationGroup & { id: string })[];
    group: Partial<(IFavoriteRegistrationGroup & { id: string })>,
    onNameChanged: (name: string) => void,
}) => {

    const [overwriteExistingGroup, setOverwriteExistingGroup] = useState(false);
    const [newId, setNewId] = useState(id);

    const handleIdChanged = useCallback((id: string) => {
        const existingGroup = groups.find((group) => group.id === id);
        if (existingGroup) {
            onNameChanged(existingGroup.name);
        }
        setNewId(id);
    }, [groups]);

    return (
        <Form style={{ paddingBottom: 0, paddingTop: "1em" }}>
            <FlexGroup className="row">
                <FormField>
                    {groups && overwriteExistingGroup
                        ? (
                            <FavoriteGroupSelect
                                onChange={handleIdChanged}
                                value={newId}
                            />
                        ) : (
                            <TextField
                                autoFocus={true}
                                outlined
                                label="Name"
                                value={name}
                                onChange={(e: React.ChangeEvent<HTMLInputElement>) => onNameChanged(e.target.value)}
                            />

                        )}
                </FormField>
                <FormField
                    first={false}
                >
                    <Checkbox
                        label="Overwrite existing"
                        checked={overwriteExistingGroup}
                        onChange={(e: React.ChangeEvent<HTMLInputElement>) => {
                            setOverwriteExistingGroup(e.target.checked);
                        }}
                    >

                    </Checkbox>
                </FormField>

            </FlexGroup>
        </Form>
    );
};
