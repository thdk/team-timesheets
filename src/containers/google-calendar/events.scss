.data-row__column.events__time {
    align-items: flex-end;
    font-size: 0.8em;
    width: 120px;
    flex-shrink: 0;
}