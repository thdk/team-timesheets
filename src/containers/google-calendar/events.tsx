import React from "react"
import { DataRow } from "../../components/data-row";


export const GoogleCalendarEvents = ({
    events,
}: {
    events: gapi.client.calendar.Event[],
}) => {
    return (
        <div>
            {events.map((event) => {
                return (
                    <DataRow
                        key={event.id}
                        line1={event.summary}
                        line2={event.description}
                        icon={"add_circle_outline"}
                    />
                );
            })}
        </div>
    );
};