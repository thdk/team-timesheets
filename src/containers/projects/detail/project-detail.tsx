import React, { useCallback } from "react";
import { TextField } from "@rmwc/textfield";

import { Form, FormField } from "../../../components/layout/form";
import { FlexGroup } from "../../../components/layout/flex";
import UserSelect from "../../users/select";
import IconPicker from "../../../components/icon-picker";

type Props = {
    readonly onNameChanged: (name: string) => void;
    readonly onIconChanged: (icon: string) => void;
    readonly onOwnerChanged: (userId: string) => void;
    readonly icon?: string;
    readonly name?: string;
    readonly ownerId?: string;
}

const ProjectDetail = (props: Props) => {

    const {
        icon,
        name,
        ownerId,
        onOwnerChanged,
        onIconChanged,
        onNameChanged,
    } = props;

    const handleNameChanged = useCallback((e: React.ChangeEvent<HTMLInputElement>) => {
        onNameChanged(e.target.value);
    }, [onNameChanged]);

    const handleIconChanged = useCallback((icon: string) => {
        onIconChanged(icon);
    }, [onIconChanged]);

    return <>
        <Form>
            <FlexGroup className="row">
                <FormField>
                    <TextField
                        outlined
                        label="Name"
                        value={name}
                        onChange={handleNameChanged}
                    />
                </FormField>
                <FormField first={false}>
                    <IconPicker
                        onChange={handleIconChanged}
                        value={icon}
                        outlined
                    />
                </FormField>
            </FlexGroup>
            <FlexGroup className="row">
                <FormField>
                    <UserSelect
                        label={"Project owner"}
                        value={ownerId}
                        onChange={onOwnerChanged}
                    />
                </FormField>
            </FlexGroup>
        </Form>
    </>;
};

export default ProjectDetail;