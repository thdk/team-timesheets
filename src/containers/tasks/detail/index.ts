export { TaskDetailContainer as TaskDetail } from "./task-detail-container";
