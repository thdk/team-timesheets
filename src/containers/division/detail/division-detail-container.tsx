import React, { useCallback } from "react";

import { DivisionDetail } from "./division-detail";
import { useDivisionStore } from "../../../contexts/division-context";
import { observer } from "mobx-react-lite";

export const DivisionDetailContainer = observer(() => {
    const divisionStore = useDivisionStore();
    const activeDocument = divisionStore.activeDocument;

    const handleIconChanged = useCallback((icon: string) => {
        if (activeDocument) {
            activeDocument.icon = icon;
        }
    }, [activeDocument]);

    const handleNameChanged = useCallback((name: string) => {
        if (activeDocument) {
            activeDocument.name = name;
        }
    }, [activeDocument]);

    return activeDocument
        ? <DivisionDetail
            onIconChanged={handleIconChanged}
            onNameChanged={handleNameChanged}
            {...activeDocument}
        />
        : null;
});
