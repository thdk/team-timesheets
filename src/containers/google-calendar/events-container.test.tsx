import React from "react";
import { render, waitFor, fireEvent } from "@testing-library/react";
import { GoogleCalendarEvents } from "./";
import { initTestFirestore, deleteFirebaseAppsAsync } from "../../__tests__/utils/firebase";
import { Store } from "../../stores/root-store";
import { events } from "./events.test";
import { IntlProvider } from "react-intl";
import { goToNewRegistration } from "../../routes/registrations/detail";
import { useGapi } from "../../hooks/use-gapi";

jest.mock("../../hooks/use-gapi");
jest.mock("../configs/use-google-config", () => ({
    useGoogleConfig: jest.fn(),
}));

jest.mock("../../routes/registrations/detail");

const {
    firestore,
} = initTestFirestore("events");

const store = new Store({
    firestore,
});

jest.mock("../../contexts/store-context", () => ({
    useStore: () => store,
}));

const getEventsList = jest.fn();

beforeAll(() => {
    (window["gapi"] as any) = {
        client: {
            calendar: {
                events: {
                    list: getEventsList
                }
            }
        }
    };

    store.user.setUser({
        uid: "user-1",
        displayName: "user 1",
        email: "email@email.com",
    } as firebase.User);

    return store.config.tasksCollection.addAsync({
        name: "Meeting",
        icon: "people",
    });
});

afterAll(() => {
    store.dispose();
    return Promise.all([
        deleteFirebaseAppsAsync(),
    ])
});

describe("GoogleCalendarEventsContainer", () => {
    it("should render without google calendar events", async () => {
        getEventsList.mockResolvedValue({
            result: {
                items: [],
            },
        });
        const { asFragment } = render(<GoogleCalendarEvents />);

        await waitFor(() => expect(asFragment()).toMatchSnapshot());

    });

    it("should render when gapi is not loaded", async () => {
        (useGapi as jest.Mock<Partial<ReturnType<typeof useGapi>>>)
            .mockReturnValue({
                user: {} as gapi.auth2.GoogleUser,
                isGapiLoaded: false,
            });

        const { asFragment } = render(<GoogleCalendarEvents />);

        await waitFor(() => expect(asFragment()).toMatchSnapshot());

        (useGapi as jest.Mock<Partial<ReturnType<typeof useGapi>>>)
            .mockReturnValue({
                user: {} as gapi.auth2.GoogleUser,
                isGapiLoaded: true,
            });
    });

    it("should render when gapi is loaded but auth user is undefined", async () => {
        (useGapi as jest.Mock<Partial<ReturnType<typeof useGapi>>>)
            .mockReturnValue({
                isGapiLoaded: true,
            });

        const { asFragment } = render(<GoogleCalendarEvents />);

        await waitFor(() => expect(asFragment()).toMatchSnapshot());

        (useGapi as jest.Mock<Partial<ReturnType<typeof useGapi>>>)
            .mockReturnValue({
                user: {} as gapi.auth2.GoogleUser,
                isGapiLoaded: true,
            });
    });

    it("should render with google calendar events", async () => {
        getEventsList.mockResolvedValue({
            result: {
                items: events,
            },
        });

        const { asFragment } = render(
            <IntlProvider
                locale={"en-US"}
                timeZone={"Europe/Brussels"}
            ><GoogleCalendarEvents />
            </IntlProvider>
        );

        await waitFor(() => expect(asFragment()).toMatchSnapshot());

    });

    it("should set selected registration with data from event on event click", async () => {
        getEventsList.mockResolvedValue({
            result: {
                items: events,
            },
        });

        const { findByText } = render(
            <IntlProvider
                locale={"en-US"}
                timeZone={"Europe/Brussels"}
            ><GoogleCalendarEvents />
            </IntlProvider>
        );

        const eventItem1 = await findByText("Summary 1");
        fireEvent.click(eventItem1);

        await waitFor(
            () => expect(
                store.timesheets.activeDocument?.sourceId).toBe("event-1")
        );

        expect(goToNewRegistration).toBeCalledTimes(1);

        const eventItem2 = await findByText("Summary 2");
        fireEvent.click(eventItem2);

        await waitFor(
            () => expect(
                store.timesheets.activeDocument
            ).toEqual(
                expect.objectContaining({
                    sourceId: "event-2",
                    time: 2,
                    description: "Summary 2",
                })
            )
        );

        expect(goToNewRegistration).toBeCalledTimes(2);
    });
});
