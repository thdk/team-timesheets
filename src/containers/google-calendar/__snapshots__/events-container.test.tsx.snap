// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`GoogleCalendarEvents should display events 1`] = `
<DocumentFragment>
  <div>
    <li
      class="mdc-list-divider"
      role="separator"
    />
    <div
      class="data-row"
    >
      <div
        class="data-row__header"
      >
        <i
          class="rmwc-icon rmwc-icon--ligature material-icons"
          title="Google Calendar"
        >
          event
        </i>
      </div>
      <div
        class="data-row__column"
        style="flex-grow: 1;"
      >
        <span
          class="data-row__line1 mdc-typography--subtitle1"
        >
          Summary 1
        </span>
        <span
          class="data-row__line2 mdc-typography--subtitle2"
        >
          All day event 1
        </span>
      </div>
      <div
        class="data-row__column events__time"
      />
    </div>
    <div
      class="data-row"
    >
      <div
        class="data-row__header"
      >
        <i
          class="rmwc-icon rmwc-icon--ligature material-icons"
          title="Google Calendar"
        >
          event
        </i>
      </div>
      <div
        class="data-row__column"
        style="flex-grow: 1;"
      >
        <span
          class="data-row__line1 mdc-typography--subtitle1"
        >
          Summary 2
        </span>
      </div>
      <div
        class="data-row__column events__time"
      >
        1:00 PM - 3:00 PM
      </div>
    </div>
    <li
      class="mdc-list-divider"
      role="separator"
    />
  </div>
</DocumentFragment>
`;

exports[`GoogleCalendarEvents should render without events 1`] = `<DocumentFragment />`;

exports[`GoogleCalendarEventsContainer should render when gapi is loaded but auth user is undefined 1`] = `<DocumentFragment />`;

exports[`GoogleCalendarEventsContainer should render when gapi is not loaded 1`] = `<DocumentFragment />`;

exports[`GoogleCalendarEventsContainer should render with google calendar events 1`] = `<DocumentFragment />`;

exports[`GoogleCalendarEventsContainer should render with google calendar events 2`] = `
<DocumentFragment>
  <div
    style="padding-left: 1em; margin-top: 2em;"
  >
    <h3
      class="mdc-typography--subtitle2"
      style="font-weight: 400;"
    >
      Suggestions for this timesheet:
    </h3>
  </div>
  <div>
    <li
      class="mdc-list-divider"
      role="separator"
    />
    <div
      class="data-row"
    >
      <div
        class="data-row__header"
      >
        <i
          class="rmwc-icon rmwc-icon--ligature material-icons"
          title="Google Calendar"
        >
          event
        </i>
      </div>
      <div
        class="data-row__column"
        style="flex-grow: 1;"
      >
        <span
          class="data-row__line1 mdc-typography--subtitle1"
        >
          Summary 1
        </span>
        <span
          class="data-row__line2 mdc-typography--subtitle2"
        >
          All day event 1
        </span>
      </div>
      <div
        class="data-row__column events__time"
      />
    </div>
    <div
      class="data-row"
    >
      <div
        class="data-row__header"
      >
        <i
          class="rmwc-icon rmwc-icon--ligature material-icons"
          title="Google Calendar"
        >
          event
        </i>
      </div>
      <div
        class="data-row__column"
        style="flex-grow: 1;"
      >
        <span
          class="data-row__line1 mdc-typography--subtitle1"
        >
          Summary 2
        </span>
      </div>
      <div
        class="data-row__column events__time"
      >
        1:00 PM - 3:00 PM
      </div>
    </div>
    <li
      class="mdc-list-divider"
      role="separator"
    />
  </div>
</DocumentFragment>
`;

exports[`GoogleCalendarEventsContainer should render without google calendar events 1`] = `<DocumentFragment />`;
